src/main/java/com/example/planit/engine/CalendarEngine.java

```diff
@@ -1402,22 +1402,69 @@ public static boolean hasAuthorizeScopesStillValid(String access_token) throws I
+    public DTOstudyPlanAndupComingStudySessionResponseToController getUserLatestStudyPlan(String sub) {
 
         try {
 
             Optional<User> maybeUser = userRepo.findUserBySubjectID(sub);
 
             if (maybeUser.isEmpty()) {
-                return new DTOstudyPlanResponseToController(false, ERROR_USER_NOT_FOUND, HttpStatus.BAD_REQUEST);
+                return new DTOstudyPlanAndupComingStudySessionResponseToController(false, ERROR_USER_NOT_FOUND, HttpStatus.BAD_REQUEST);
             }
-            return new DTOstudyPlanResponseToController(true, NO_PROBLEM, HttpStatus.OK, maybeUser.get().getLatestStudyPlan());
+            User user = maybeUser.get();
+            // refresh access_token before making api call
+            validateAccessTokenExpireTime(user);
+
+            // check if access_token still have scope for Google calendar
+            if (!hasAuthorizeScopesStillValid(user.getAuth().getAccessToken())) {
+                return new DTOstudyPlanAndupComingStudySessionResponseToController(false, Constants.ERROR_NO_VALID_ACCESS_TOKEN, HttpStatus.UNAUTHORIZED);
+            }
+
+            return new DTOstudyPlanAndupComingStudySessionResponseToController(true, NO_PROBLEM, HttpStatus.OK, maybeUser.get().getLatestStudyPlan(), getUpComingSessionForUser(user));
         } catch (Exception e) {
             logger.error(buildExceptionMessage(e));
-            return new DTOstudyPlanResponseToController(false, ERROR_DEFAULT, HttpStatus.INTERNAL_SERVER_ERROR);
+            return new DTOstudyPlanAndupComingStudySessionResponseToController(false, ERROR_DEFAULT, HttpStatus.INTERNAL_SERVER_ERROR);
         }
     }
 
+    private StudySession getUpComingSessionForUser(User user) throws GeneralSecurityException, IOException {
+
+        Calendar calendarService = getCalendarService(user.getAuth().getAccessToken(), user.getAuth().getExpireTimeInMilliseconds());
+
+        List<CalendarListEntry> calendarList = getCalendarList(calendarService);
+
+        StudySession upComingSession = null;
+
+        Instant currentTime = Instant.now();
+        DateTime startTime = new DateTime(currentTime.toEpochMilli());
+
+        int countEvents = 0;
+        Events events = null;
+
+        for (CalendarListEntry calendar : calendarList) {
+            if (calendar.getId().equals(user.getPlanItCalendarID())) {
+                events = calendarService.events().list(calendar.getId())
+                        .setTimeMin(startTime)
+                        .setOrderBy("startTime")
+                        .setSingleEvents(true)
+                        .execute();
+            }
+        }
+
+        // if the user have calendar. we return the upComing study session.
+        if (events != null) {
+            List<Event> ListOfEvent = events.getItems();
+            if (ListOfEvent.size() != 0)
+            {
+                Event event = ListOfEvent.get(countEvents);
```

Comment:
@oshriFadlon1 As I see, the `countEvents` is always 0. Then why is it a variable?
Maybe you can add `final` if you want to give it a meaningful name, but then the name "countEvents" does not make sence. because it does not represent the count of the events.
So I think you meant to give it a meaningful name and make if `final` or get rid of the variable and just put "0" there.

Reply:
fixed